ResearchKit/Common/ORKPieChartView.m

```diff
@@ -0,0 +1,543 @@
+    cell.titleLabel.text = [_datasource pieChartView:self titleForSegmentAtIndex:indexPath.item];
+    cell.titleLabel.font = _legendFont;
+    cell.dotView.backgroundColor = [_datasource pieChartView:self colorForSegmentAtIndex:indexPath.item];
+    cell.transform = CGAffineTransformMakeScale(0, 0);
+    return cell;
+}
+
+- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
+    return [_datasource numberOfSegmentsInPieChartView];
+}
+
+- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
+    ORKLegendCollectionViewCell *c = [[ORKLegendCollectionViewCell alloc]initWithFrame:CGRectZero];
+    c.titleLabel.text = [_datasource pieChartView:self titleForSegmentAtIndex:indexPath.item];
+    c.titleLabel.font = _legendFont;
+    [c.contentView setNeedsUpdateConstraints];
+    CGSize s = [c.contentView systemLayoutSizeFittingSize:UILayoutFittingCompressedSize];
+    return s;
+}
+
+- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section {
+    return 0;
+}
+
+- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section {
+    return 15;
+}
+
+#pragma mark - DataSource
+
+- (NSInteger)numberOfSegments {
+    NSInteger count = 0;
+    if ([self.datasource respondsToSelector:@selector(numberOfSegmentsInPieChartView)]) {
+        count = [self.datasource numberOfSegmentsInPieChartView];
+    }
+    return count;
+}
+
+- (UIColor *)colorForSegmentAtIndex:(NSInteger)index {
+    UIColor *color = nil;
+    if ([self.datasource respondsToSelector:@selector(pieChartView:colorForSegmentAtIndex:)]) {
+        color = [self.datasource pieChartView:self colorForSegmentAtIndex:index];
+    }
+    else {
+        // Default colors
+        NSInteger numberOfSegments = [self numberOfSegments];
+        if (numberOfSegments > 1) {
+            CGFloat divisionFactor = (CGFloat)(1/(CGFloat)(numberOfSegments -1));
+            color = [UIColor colorWithWhite:(divisionFactor * index) alpha:1.0f];
+        }
+        else {
+            color = [UIColor grayColor];
+        }
+    }
+    return color;
+}
+
+- (CGFloat)valueForSegmentAtIndex:(NSInteger)index {
+    CGFloat value = 0;
+    if ([self.datasource respondsToSelector:@selector(pieChartView:valueForSegmentAtIndex:)]) {
+        value = [self.datasource pieChartView:self valueForSegmentAtIndex:index];
+    }
+    return value;
+}
+
+#pragma mark - Draw
+
+- (void)drawLegend {
+    if (self.shouldAnimate) {
+        NSArray *sorted = [_legendView.visibleCells sortedArrayUsingComparator:^NSComparisonResult(UICollectionViewCell *cell1, UICollectionViewCell *cell2) {
+            return cell1.frame.origin.x > cell2.frame.origin.x;
```

Comment:
I'd happy to just have the simpler case of legend cells sorted at their original collection view sort order (i.e., first row would draw first than second row).

But this solution is not bad either, however, we'd need to check for `origin.y` where two cells have the same `origin.x` (higher cell would be earlier than lower cell).